src/groupeddataframe/groupeddataframe.jl

```diff
@@ -245,6 +251,106 @@ Base.IndexStyle(::Type{<:GroupKeys}) = IndexLinear()
 end
 
 
+#
+# Non-standard indexing
+#
+
+# Non-standard indexing relies on converting to integer indices first
+# The full version (to_indices) is required rather than to_index even though
+# GroupedDataFrame behaves as a 1D array due to the behavior of Colon and Not.
+# Note that this behavior would be the default if it was <:AbstractArray
+Base.getindex(gd::GroupedDataFrame, idx...) = getindex(gd, Base.to_indices(gd, idx)...)
+
+# The allowed key types for dictionary-like indexing
+const GroupKeyTypes = Union{GroupKey, Tuple, NamedTuple}
+# All allowed scalar index types
+const GroupIndexTypes = Union{Integer, GroupKeyTypes}
+
+# Find integer index for dictionary keys:
+function Base.to_index(gd::GroupedDataFrame, key::GroupKey)
+    gd === parent(key) && return getfield(key, :idx)
+    throw(ErrorException("Cannot use a GroupKey to index a GroupedDataFrame other than the one it was derived from."))
+end
+
+function Base.to_index(gd::GroupedDataFrame, key::Tuple)
+    for i in 1:length(gd)
+        isequal(Tuple(_groupvalues(gd, i)), key) && return i
+    end
+    throw(KeyError(key))
+end
+
+function Base.to_index(gd::GroupedDataFrame, key::NamedTuple{N}) where {N}
+    if length(key) != length(gd.cols) || any(n != _names(gd)[c] for (n, c) in zip(N, gd.cols))
+        throw(KeyError(key))
+    end
+    return Base.to_index(gd, Tuple(key))
+end
+
+# Array of (possibly non-standard) indices
+function Base.to_index(gd::GroupedDataFrame, idxs::AbstractVector{T}) where {T}
+    # A concrete eltype which is <: GroupIndexTypes, don't need to check
+    isconcretetype(T) && T <: GroupIndexTypes && return [Base.to_index(gd, i) for i in idxs]
+
+    # Edge case - array is empty
+    isempty(idxs) && return Int[]
+
+    # Infer eltype based on type of first index
+    idx1 = idxs[1]
+    E1 = typeof(idx1)
+
+    E = if E1 <: Integer && E1 !== Bool
+        Integer
+    elseif E1 <: GroupKey
```

Comment:
In Base integer indices are allowed to have mixed types as long as they are not `Bool` if mixed types are present.

Reply:
OK. So there's no simpler way to achieve this?

Reply:
We do not have to be super elegant here (and roughly it is a minimal way to implement it - I have not thought much if it can be improved).
This method is just for completeness, as it will be hardly ever called (it will be used only if the users passes a vector that has an abstract eltype).